File: haskell/env.hs
module Env
  ( State
  , Action
  , Reward
  , initState
  , oneStep
  , reward
  )
where

import           Data.Fixed

g = 9.80665 -- 重力加速度
mCart = 1.0 -- カートの質量
mPole = 0.1 -- ポールの質量
l = 0.5 -- ポールの半分の長さ
fps = 50 -- frames per second
tau = 1 / fps -- 制御周期

-- あとでつかう
ml = mPole * l
mass = mCart + mPole


-- 状態ベクトル
-- 最初 tuple で実装しようと思ったけど気が遠くなるので List にした
type State = [Double]

-- 状態ベクトルの初期値
initState :: State
initState = [0.0, -pi, 0.0, 0.0]


-- 行動の値
type Action = Double


-- 報酬の型
type Reward = Double


reward :: State -> Double
reward s | abs x > 2.0 = -2.0
         | otherwise   = -abs theta + pi / 2
 where
  x     = s !! 0
  theta = s !! 1

oneStep :: State -> Action -> State
oneStep s a = rungeKuttaSolve s a tau

-- 状態 s で力 u を加えたときの微分
differential :: State -> Action -> State
differential s u = [xdot, thetadot, xddot, thetaddot]
 where
  theta    = s !! 1
  xdot     = s !! 2
  thetadot = s !! 3
  sintheta = sin theta
  costheta = cos theta
  -- なんじゃこりゃー（hie が整形した）
  xddot =
    (  4
      *  u
      /  3
      +  4
      *  ml
      *  thetadot
      ** 2
      *  sintheta
      /  3
      -  mPole
      *  g
      *  sin (2 * theta)
      /  2
      )
      / (4 * mass - mPole * costheta ** 2)
  thetaddot =
    (  mass
      *  g
      *  sintheta
      -  ml
      *  thetadot
      ** 2
      *  sintheta
      *  costheta
      -  u
      *  costheta
      )
      / (4 * mass * l / 3 - ml * costheta ** 2)


-- オイラー法で微分方程式を解く
eulerSolve :: State -> State -> Double -> State
eulerSolve s sdot dt = map (\ssdot -> fst ssdot + snd ssdot * dt) $ zip s sdot


-- ルンゲクッタ法で微分方程式を解く
rungeKuttaSolve :: State -> Double -> Double -> State
rungeKuttaSolve s u dt =
  [ snext !! 0
  , mod' ((snext !! 1) + 3.0 * pi) 2 * pi - pi
  , snext !! 2
  , snext !! 3
  ]
 where
  k1 = differential s u
  s1 = eulerSolve s k1 (dt / 2)
  k2 = differential s1 u
  s2 = eulerSolve s k2 (dt / 2)
  k3 = differential s2 u
  s3 = eulerSolve s k3 dt
  k4 = differential s3 u
  -- つらい
  snext =
    [ s !! 0 + (k1 !! 0 + 2 * k2 !! 0 + 2 * k3 !! 0 + k4 !! 0) * dt / 6
    , s !! 1 + (k1 !! 1 + 2 * k2 !! 1 + 2 * k3 !! 1 + k4 !! 1) * dt / 6
    , s !! 2 + (k1 !! 2 + 2 * k2 !! 2 + 2 * k3 !! 2 + k4 !! 2) * dt / 6
    , s !! 3 + (k1 !! 3 + 2 * k2 !! 3 + 2 * k3 !! 3 + k4 !! 3) * dt / 6
    ]

File: rust/src/env.rs
use std::f64::consts::PI;

const G: f64 = 9.80665; // 重力加速度
const MCART: f64 = 1.0; // カートの質量
const MPOLE: f64 = 0.1; // ポールの質量
const L: f64 = 0.5; // ポールの半分の長さ
pub const FPS: u32 = 50; // frames per second
const TAU: f64 = 1.0 / FPS as f64; // 制御周期

// あとの計算でつかう
const ML: f64 = MPOLE * L;
const MASS: f64 = MCART + MPOLE;

const STATE_LEN: usize = 4; // 状態ベクトルの長さ

pub type State = [f64; 4];
pub type Action = f64;
pub type Reward = f64;

pub fn new_state() -> State {
    // [x, theta, xdot, thetadot]
    [0.0, -PI, 0.0, 0.0]
}

pub fn step(s: &State, a: Action) -> State {
    runge_kutta_solve(s, a, TAU)
}

pub fn reward(s: &State) -> Reward {
    let [x, theta, _, _] = s;
    if x.abs() > 2.0 {
        return -2.0;
    }
    -theta.abs() + PI / 2.0
}

/// 状態 s で力 u を加えたときの微分
fn differential(s: &State, u: Action) -> State {
    let [_x, theta, xdot, thetadot] = s;
    let sintheta = theta.sin();
    let costheta = theta.cos();

    let xddot = (4.0 * u / 3.0 + 4.0 * ML * thetadot.powf(2.0) * sintheta / 3.0
        - MPOLE * G * (2.0 * theta).sin() / 2.0)
        / (4.0 * MASS - MPOLE * costheta.powf(2.0));
    let thetaddot =
        (MASS * G * sintheta - ML * thetadot.powf(2.0) * sintheta * costheta - u * costheta)
            / (4.0 * MASS * L / 3.0 - ML * costheta.powf(2.0));

    [*xdot, *thetadot, xddot, thetaddot]
}

/// オイラー法を用いて微分方程式を解く
fn euler_solve(s: &State, sdot: &State, dt: f64) -> State {
    let mut ans = [0.0; STATE_LEN];
    for i in 0..STATE_LEN {
        ans[i] = s[i] + sdot[i] * dt;
    }
    ans
}

fn runge_kutta_solve(s: &State, u: Action, dt: f64) -> State {
    let k1 = differential(s, u);
    let s1 = euler_solve(s, &k1, dt / 2.0);
    let k2 = differential(&s1, u);
    let s2 = euler_solve(s, &k2, dt / 2.0);
    let k3 = differential(&s2, u);
    let s3 = euler_solve(s, &k3, dt);
    let k4 = differential(&s3, u);

    let mut snext = [0.0; STATE_LEN];
    for i in 0..STATE_LEN {
        snext[i] = s[i] + (k1[i] + 2.0 * k2[i] + 2.0 * k3[i] + k4[i]) * dt / 6.0;
    }

    snext[1] = (snext[1] + 3.0 * PI) % (2.0 * PI) - PI;
    snext
}
